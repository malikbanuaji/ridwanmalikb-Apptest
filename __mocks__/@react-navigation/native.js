export const useRoute = jest.fn(() => ({
  params: {},
}));
export const useNavigation = () => ({
  navigate: jest.fn(),
  goBack: jest.fn(),
});
