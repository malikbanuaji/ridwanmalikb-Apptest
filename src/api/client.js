import axios from 'axios';

const client = axios.create({
  baseURL: 'https://simple-contact-crud.herokuapp.com/',
  timeout: 3000,
});

const getAllContact = () => client.get('/contact');

const addContact = payload => client.post('/contact', payload);

const getContact = id => client.get(`/contact/${id}`);

const updateContact = (id, payload) => client.put(`/contact/${id}`, payload);

const deleteContact = (id, payload) => client.delete(`/contact/${id}`, payload);

export const contactAPI = {
  getAllContact,
  addContact,
  getContact,
  updateContact,
  deleteContact,
};
