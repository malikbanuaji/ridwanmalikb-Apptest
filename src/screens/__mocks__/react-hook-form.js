import * as React from 'react';

export const useForm = jest.fn();

const ControllerComponent = 'Controller';
export const Controller = props => {
  const {render, ...restProps} = props;
  return (
    <ControllerComponent {...restProps}>
      {render({field: {onChange: jest.fn(), value: jest.fn()}})}
    </ControllerComponent>
  );
};
