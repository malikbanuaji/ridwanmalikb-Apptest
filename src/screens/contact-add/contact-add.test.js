import * as React from 'react';

import {render, waitFor, cleanup} from '@testing-library/react-native';
import ContactAdd from './contact-add';
import {Alert, ToastAndroid} from 'react-native';
import {useRoute} from '@react-navigation/native';
import Constants from '../../constants/constants';

import {useForm} from 'react-hook-form';
import {act} from 'react-test-renderer';
import {contactAPI} from '../../api/client';

jest
  .mock('./hooks/useContactAdd', () => jest.fn())
  .mock('./components/text-input-controller', () => 'TextInputController')
  .mock('../../components')
  .mock('react-redux', () => ({
    useDispatch: jest.fn(() => param => param),
  }));

const findByButtonTitle = (jsonComponent = [], title) => {
  if (jsonComponent.type === 'Button' && jsonComponent.props.title === title) {
    return jsonComponent;
  } else {
    const children = jsonComponent.children ? jsonComponent.children : [];

    for (let i = 0; i < children.length; i += 1) {
      const currentChild = jsonComponent.children[i];
      const result = findByButtonTitle(currentChild, title);
      if (result !== false) {
        return result;
      }
    }
    return false;
  }
};

describe('ContactAdd test', () => {
  const navigation = {
    goBack: jest.fn(),
  };
  const useFormMock = {
    watch: () => ({
      firstName: 'malik',
      lastName: 'malik',
    }),
    control: 'control',
    handleSubmit: jest.fn(cb => cb),
    reset: jest.fn(),
    setValue: jest.fn(),
    formState: {errors: {}, isValid: false},
  };

  const toastAndroidShow = jest
    .spyOn(ToastAndroid, 'show')
    .mockReturnValue(undefined);

  beforeEach(() => {
    useForm.mockImplementation(() => useFormMock);
    jest
      .spyOn(contactAPI, 'addContact')
      .mockImplementationOnce(() => Promise.resolve());
    jest
      .spyOn(contactAPI, 'updateContact')
      .mockImplementationOnce(() => Promise.resolve({}));
    jest
      .spyOn(contactAPI, 'deleteContact')
      .mockImplementationOnce(() => Promise.resolve({}));
  });

  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });

  test('should render properly with add type', () => {
    useRoute.mockImplementationOnce(() => ({
      params: {
        id: '1',
        type: Constants.CONTACT_ADD_ACTION_TYPE.ADD,
      },
    }));
    const component = render(<ContactAdd navigation={navigation} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render properly with view type', () => {
    useRoute.mockImplementationOnce(() => ({
      params: {
        id: '1',
        type: Constants.CONTACT_ADD_ACTION_TYPE.VIEW,
      },
    }));
    const component = render(<ContactAdd navigation={navigation} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render properly with Save Contact', async () => {
    const routeMock = {
      params: {
        id: '1',
        type: Constants.CONTACT_ADD_ACTION_TYPE.VIEW,
      },
    };
    useRoute.mockImplementationOnce(() => routeMock);
    jest.spyOn(React, 'useState').mockImplementationOnce();

    const component = render(<ContactAdd navigation={navigation} />);
    const data = {
      firstName: 'First',
    };
    await act(async () =>
      findByButtonTitle(component.toJSON(), 'Save Contact').props.onPress(data),
    );

    await waitFor(async () => {
      expect(contactAPI.updateContact).toBeCalledWith(routeMock.params.id, {
        ...data,
        photo: 'N/A',
      });
    });

    expect(toastAndroidShow).toBeCalled();
  });

  test('should render properly with Add Contact', async () => {
    const routeMock = {
      params: {
        id: '1',
        type: Constants.CONTACT_ADD_ACTION_TYPE.ADD,
      },
    };
    useRoute.mockImplementationOnce(() => routeMock);
    jest.spyOn(React, 'useState').mockImplementationOnce();

    const component = render(<ContactAdd navigation={navigation} />);
    const data = {
      firstName: 'First',
    };
    await act(async () =>
      findByButtonTitle(component.toJSON(), 'Add Contact').props.onPress(data),
    );

    await waitFor(async () => {
      expect(contactAPI.addContact).toBeCalledWith({
        ...data,
        photo: 'N/A',
      });
    });

    expect(toastAndroidShow).toBeCalled();
  });

  test('should render properly with Delete Contact', async () => {
    const routeMock = {
      params: {
        id: '1',
        type: Constants.CONTACT_ADD_ACTION_TYPE.VIEW,
      },
    };
    useRoute.mockImplementationOnce(() => routeMock);
    jest.spyOn(React, 'useState').mockImplementationOnce();

    const Alertmock = jest
      .spyOn(Alert, 'alert')
      .mockImplementation(() => jest.fn());

    const component = render(<ContactAdd navigation={navigation} />);

    findByButtonTitle(component.toJSON(), 'Delete Contact').props.onPress();
    expect(Alertmock).toBeCalled();
    await act(async () => Alertmock.mock.calls[0][2][1].onPress());
    await waitFor(async () => {
      expect(contactAPI.deleteContact).toBeCalledWith(routeMock.params.id);
    });

    expect(toastAndroidShow).toBeCalled();
  });
});

describe('ContactAdd test negative', () => {
  const navigation = {
    goBack: jest.fn(),
  };
  const useFormMock = {
    watch: () => ({
      firstName: 'malik',
      lastName: 'malik',
    }),
    control: 'control',
    handleSubmit: jest.fn(cb => cb),
    reset: jest.fn(),
    setValue: jest.fn(),
    formState: {errors: {}, isValid: false},
  };

  const toastAndroidShow = jest
    .spyOn(ToastAndroid, 'show')
    .mockReturnValue(undefined);
    const error = new Error('error');

  const addContactApi = jest
    .spyOn(contactAPI, 'addContact')
    .mockRejectedValueOnce(error);
  jest.spyOn(contactAPI, 'updateContact').mockRejectedValueOnce(error);
  jest.spyOn(contactAPI, 'deleteContact').mockRejectedValueOnce(error);

  beforeEach(() => {
    useForm.mockImplementation(() => useFormMock);

  });

  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });

  test('should render properly with Save Contact', async () => {
    const routeMock = {
      params: {
        id: '1',
        type: Constants.CONTACT_ADD_ACTION_TYPE.VIEW,
      },
    };
    useRoute.mockImplementationOnce(() => routeMock);
    jest.spyOn(React, 'useState').mockImplementationOnce();

    const component = render(<ContactAdd navigation={navigation} />);
    const data = {
      firstName: 'First',
    };
    await act(async () =>
      findByButtonTitle(component.toJSON(), 'Save Contact').props.onPress(data),
    );

    await waitFor(async () => {
      expect(contactAPI.updateContact).toBeCalledWith(routeMock.params.id, {
        ...data,
        photo: 'N/A',
      });
    });

    expect(toastAndroidShow).toBeCalled();
  });

  test('should render properly with Add Contact', async () => {
    const routeMock = {
      params: {
        id: '1',
        type: Constants.CONTACT_ADD_ACTION_TYPE.ADD,
      },
    };
    useRoute.mockImplementationOnce(() => routeMock);
    jest.spyOn(React, 'useState').mockImplementationOnce();

    const component = render(<ContactAdd navigation={navigation} />);
    const data = {
      firstName: 'First',
    };
    await act(async () =>
      findByButtonTitle(component.toJSON(), 'Add Contact').props.onPress(data),
    );

    await waitFor(async () => {
      expect(addContactApi).rejects.toEqual('errorr');
    });

    expect(toastAndroidShow).toBeCalled();
  });

  test('should render properly with Delete Contact', async () => {
    const routeMock = {
      params: {
        id: '1',
        type: Constants.CONTACT_ADD_ACTION_TYPE.VIEW,
      },
    };
    useRoute.mockImplementationOnce(() => routeMock);
    jest.spyOn(React, 'useState').mockImplementationOnce();

    const Alertmock = jest
      .spyOn(Alert, 'alert')
      .mockImplementation(() => jest.fn());

    const component = render(<ContactAdd navigation={navigation} />);

    findByButtonTitle(component.toJSON(), 'Delete Contact').props.onPress();
    expect(Alertmock).toBeCalled();
    await act(async () => Alertmock.mock.calls[0][2][1].onPress());
    await waitFor(async () => {
      expect(contactAPI.deleteContact).toBeCalledWith(routeMock.params.id);
    });

    expect(toastAndroidShow).toBeCalled();
  });
});
