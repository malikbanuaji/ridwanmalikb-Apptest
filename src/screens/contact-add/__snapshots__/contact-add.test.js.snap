// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactAdd test should render properly with add type 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#36393f",
      "flex": 1,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "backgroundColor": "#2f3136",
            "borderRadius": 8,
            "marginBottom": 20,
            "marginHorizontal": 20,
            "padding": 20,
          }
        }
      >
        <Controller
          control="control"
          name="photo"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "borderRadius": 8,
                "height": 130,
                "justifyContent": "center",
                "overflow": "hidden",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <ProfilePicture
                photoUri={[MockFunction]}
                size={130}
              />
            </View>
            <Pressable
              onPress={[Function]}
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "height": 130,
                    "justifyContent": "center",
                    "width": 130,
                  }
                }
              >
                <AntDesignIcon
                  color="rgba(255,255,255,0.5)"
                  name="edit"
                  size={50}
                />
              </View>
            </Pressable>
          </View>
        </Controller>
        <TextInputController
          control="control"
          errors={Object {}}
          label="First name"
          name="firstName"
          placeholder="e.g. John"
          rules={
            Object {
              "maxLength": Object {
                "message": "Max. length 30",
                "value": 30,
              },
              "minLength": Object {
                "message": "Min. length 3",
                "value": 3,
              },
              "pattern": Object {
                "message": "Only alpha-numeric allowed",
                "value": /\\^\\[A-Za-z0-9\\]\\+\\$/gm,
              },
              "required": "This is required",
            }
          }
        />
        <TextInputController
          control="control"
          errors={Object {}}
          label="Last name"
          name="lastName"
          placeholder="e.g. Doe"
          rules={
            Object {
              "maxLength": Object {
                "message": "Max. length 30",
                "value": 30,
              },
              "minLength": Object {
                "message": "Min. length 3",
                "value": 3,
              },
              "pattern": Object {
                "message": "Only alpha-numeric allowed",
                "value": /\\^\\[A-Za-z0-9\\]\\+\\$/gm,
              },
              "required": "This is required",
            }
          }
        />
        <TextInputController
          control="control"
          errors={Object {}}
          keyboardType="number-pad"
          label="Age"
          name="age"
          placeholder="e.g. 20"
          rules={
            Object {
              "max": Object {
                "message": "Minimum age is 100",
                "value": 100,
              },
              "min": Object {
                "message": "Minimum age is 1",
                "value": 1,
              },
              "pattern": Object {
                "message": "Must be number only",
                "value": /\\^\\[0-9\\]\\+\\$/gm,
              },
              "required": "This is required",
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "marginHorizontal": 20,
          }
        }
      >
        <Button
          disabled={true}
          onPress={[Function]}
          title="Add Contact"
          type="primary"
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`ContactAdd test should render properly with view type 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#36393f",
      "flex": 1,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "backgroundColor": "#2f3136",
            "borderRadius": 8,
            "marginBottom": 20,
            "marginHorizontal": 20,
            "padding": 20,
          }
        }
      >
        <Controller
          control="control"
          name="photo"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "borderRadius": 8,
                "height": 130,
                "justifyContent": "center",
                "overflow": "hidden",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <ProfilePicture
                photoUri={[MockFunction]}
                size={130}
              />
            </View>
            <Pressable
              onPress={[Function]}
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "height": 130,
                    "justifyContent": "center",
                    "width": 130,
                  }
                }
              >
                <AntDesignIcon
                  color="rgba(255,255,255,0.5)"
                  name="edit"
                  size={50}
                />
              </View>
            </Pressable>
          </View>
        </Controller>
        <TextInputController
          control="control"
          errors={Object {}}
          label="First name"
          name="firstName"
          placeholder="e.g. John"
          rules={
            Object {
              "maxLength": Object {
                "message": "Max. length 30",
                "value": 30,
              },
              "minLength": Object {
                "message": "Min. length 3",
                "value": 3,
              },
              "pattern": Object {
                "message": "Only alpha-numeric allowed",
                "value": /\\^\\[A-Za-z0-9\\]\\+\\$/gm,
              },
              "required": "This is required",
            }
          }
        />
        <TextInputController
          control="control"
          errors={Object {}}
          label="Last name"
          name="lastName"
          placeholder="e.g. Doe"
          rules={
            Object {
              "maxLength": Object {
                "message": "Max. length 30",
                "value": 30,
              },
              "minLength": Object {
                "message": "Min. length 3",
                "value": 3,
              },
              "pattern": Object {
                "message": "Only alpha-numeric allowed",
                "value": /\\^\\[A-Za-z0-9\\]\\+\\$/gm,
              },
              "required": "This is required",
            }
          }
        />
        <TextInputController
          control="control"
          errors={Object {}}
          keyboardType="number-pad"
          label="Age"
          name="age"
          placeholder="e.g. 20"
          rules={
            Object {
              "max": Object {
                "message": "Minimum age is 100",
                "value": 100,
              },
              "min": Object {
                "message": "Minimum age is 1",
                "value": 1,
              },
              "pattern": Object {
                "message": "Must be number only",
                "value": /\\^\\[0-9\\]\\+\\$/gm,
              },
              "required": "This is required",
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "marginHorizontal": 20,
          }
        }
      >
        <Button
          disabled={true}
          onPress={[Function]}
          title="Save Contact"
          type="primary"
        />
        <View
          style={
            Object {
              "marginTop": 10,
            }
          }
        >
          <Button
            onPress={[Function]}
            title="Delete Contact"
            type="danger"
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
